iwyu.cc

```diff
@@ -1494,8 +1497,11 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   }
 
   set<const Type*> GetProvidedTypesForFnReturn(const FunctionDecl* decl) const {
-    const Type* return_type = decl->getReturnType().getTypePtr();
-    return GetProvidedTypes(return_type, GetLocation(decl));
+    const FunctionDecl* providing_decl = decl;
+    if (const auto* method = dyn_cast<CXXMethodDecl>(decl))
+      providing_decl = GetUpmostInHierarchy(method);
```

Comment:
So in a scenario like:
```
class Base {};
class Derived0 : public Base {};
class Derived1 : public Derived0 {};
class Derived2 : public Derived1 {};
class Derived3 : public Derived2 {};
```
... we run through the whole chain up to the least derived/most base and use its return type/parameter types.

Do you think maybe we'd need to check each intermediary step too? I.e. if a type is provided at any level in the inheritance chain, it would be provided below, right? I feel like that would make this a lot more complicated, though.

Reply:
I just thought that IWYU should generally remove all `#include`s and fwd-decls for return and parameter types at any level except `Base`, hence a user cannot provide a type at the other levels. Moreover, according to the default policy, the absence of both `#include` and fwd-decl means for IWYU that the type is provided, which would be obviously wrong in many cases.

Reply:
I see, that makes sense.

> Moreover, according to the default policy, the absence of both #include and fwd-decl means for IWYU that the type is provided, which would be obviously wrong in many cases.

I thought the policy was:

* `#include`: author is responsible (i.e. provided)
* fwd decl: user is responsible (i.e. not provided)
* neither: can't compile?

so I'm not really getting that... Or wait, you mean that "is provided?" only returns true if there is a forward-decl, whether there's an `#include` directive in place or not (on the assumption that code can't compile without a declaration)?

Reply:
> neither: can't compile?

`#include` or fwd-decl can come via some other header, even via `stdafx.h`, so the code can actually compile. `CodeAuthorWantsJustAForwardDeclare` function considers only direct `#include`s and fwd-decls from the file defining a type alias or a function prototype.

Indeed, the proposed changes are to avoid intentionally the collecting of provided types from intermediate method declarations, otherwise `FwdRetType` and `FwdAutoconvParam` are considered as provided by `Derived1` and by `Derived2` just because they are not forward-declared (neither included) in `reexport_overridden-d1.h` (see [the comment](https://github.com/include-what-you-use/include-what-you-use/pull/1596/files#diff-4476e0e75321badf44d0ba46b72836581736bf2628b62859baf7d0774f9d8f28R20)). You can compare with the behavior on the current master.

Reply:
> considers only *direct* #includes and fwd-decls from the file defining a type alias or a function prototype.

Ah, that's what I was missing, thanks.

I feel like there might be a use-case for breaking the reexport chain halfway through a hierarchy, but I can't really put it into words. And the suggested behavior is so nice and simple that it feels better to find out through testing in the wild.